chain/chain/src/chain.rs

```diff
@@ -3500,972 +3891,503 @@ impl Chain {
-        });
-        new_shards
-            .iter()
-            .map(|shard_uid| {
-                self.chain_store_update
-                    .get_chunk_extra(block.header().prev_hash(), shard_uid)
-                    .map(|chunk_extra| (*shard_uid, *chunk_extra.state_root()))
+    pub fn group_receipts_by_shard(
+        receipts: Vec<Receipt>,
+        shard_layout: &ShardLayout,
+    ) -> HashMap<ShardId, Vec<Receipt>> {
+        let mut result = HashMap::with_capacity(shard_layout.num_shards() as usize);
+        for receipt in receipts {
+            let shard_id = account_id_to_shard_id(&receipt.receiver_id, shard_layout);
+            let entry = result.entry(shard_id).or_insert_with(Vec::new);
+            entry.push(receipt)
+        }
+        result
+    }
+
+    pub fn build_receipts_hashes(
+        receipts: &[Receipt],
+        shard_layout: &ShardLayout,
+    ) -> Vec<CryptoHash> {
+        if shard_layout.num_shards() == 1 {
+            return vec![hash(&ReceiptList(0, receipts).try_to_vec().unwrap())];
+        }
+        let mut account_id_to_shard_id_map = HashMap::new();
+        let mut shard_receipts: Vec<_> =
+            (0..shard_layout.num_shards()).map(|i| (i, Vec::new())).collect();
+        for receipt in receipts.iter() {
+            let shard_id = match account_id_to_shard_id_map.get(&receipt.receiver_id) {
+                Some(id) => *id,
+                None => {
+                    let id = account_id_to_shard_id(&receipt.receiver_id, shard_layout);
+                    account_id_to_shard_id_map.insert(receipt.receiver_id.clone(), id);
+                    id
+                }
+            };
+            shard_receipts[shard_id as usize].1.push(receipt);
+        }
+        shard_receipts
+            .into_iter()
+            .map(|(i, rs)| {
+                let bytes = (i, rs).try_to_vec().unwrap();
+                hash(&bytes)
             })
             .collect()
     }
+}
 
-    /// Creates jobs that would apply chunks
-    fn apply_chunks_preprocessing(
-        &mut self,
-        me: &Option<AccountId>,
-        block: &Block,
-        prev_block: &Block,
-        incoming_receipts: &HashMap<ShardId, Vec<ReceiptProof>>,
-        mode: ApplyChunksMode,
-    ) -> Result<
-        Vec<Box<dyn FnOnce(&Span) -> Result<ApplyChunkResult, Error> + Send + 'static>>,
-        Error,
-    > {
-        let _span = tracing::debug_span!(target: "chain", "apply_chunks_preprocessing").entered();
-        let mut result: Vec<
-            Box<dyn FnOnce(&Span) -> Result<ApplyChunkResult, Error> + Send + 'static>,
-        > = Vec::new();
-        #[cfg(not(feature = "mock_node"))]
-        let protocol_version =
-            self.runtime_adapter.get_epoch_protocol_version(block.header().epoch_id())?;
+/// Sandbox node specific operations
+impl Chain {
+    pub fn patch_state(&mut self, patch: StatePatch) {
```

Comment:
Should we still add #[cfg(feature = "sandbox")] here?

Reply:
I think we shouldn't. The current design guarantees that you only can have an `StatePatch` if `cfg(feature = "sandbox")` is true. That is, the only thing which is gated by the feature are `StatePatch` constructors. 

I am bit torn on whether this is a good approach. I *like* it because it is more convenient (conditional compilation is localized) and because it is more robust (you don't need to make sure that *every* use-site is feature gated, there's type-system level guarantee that, to do anything with StatePatch, you have to be sandbox). 

I *don't* like it because it is somewhat non-obvious, and any future reader would think "wait, shouldn't this be gated by cfg", and would have to spend some cycles looking at the definition of StatePatch and figuring the pattern of gating constructors out. 

On balance, I think this falls under acceptable and encourages uses of the type system, and is not an abuse just to add some fanciness. But happy to reconsider if folks feel otherwise!

Reply:
Can we add a comment in the code documenting that? I think that should be good!